app/services/s3_email_archive_service.rb

```diff
@@ -0,0 +1,53 @@
+class S3EmailArchiveService
+  def self.call(*args)
+    new.call(*args)
+  end
+
+  # For batch we expect an array of hashes containing email data in the format
+  # from EmailArchivePresenter
+  def call(batch)
+    group_by_date(batch).map { |prefix, records| send_to_s3(prefix, records) }
+  end
+
+  private_class_method :new
+
+private
+
+  def group_by_date(batch)
+    batch.group_by do |item|
+      # we group by date in this way to create partitions for s3/athena
+      # these are grouped in case dates span more than one day
+      Date.parse(
+        item.fetch(:finished_sending_at_utc)
+      ).strftime("year=%Y/month=%m/date=%d")
+    end
+  end
+
+  def send_to_s3(prefix, records)
+    records = records.sort_by { |r| r.fetch(:finished_sending_at_utc) }
+    last_time = records.last[:finished_sending_at_utc]
+    obj = bucket.object(object_name(prefix, last_time))
+    obj.put(
+      body: object_body(records),
+      content_encoding: "gzip"
+    )
+  end
+
+  def bucket
+    @bucket ||= begin
+                  s3 = Aws::S3::Resource.new
+                  s3.bucket(ENV.fetch("EMAIL_ARCHIVE_S3_BUCKET"))
+                end
+  end
+
+  def object_name(prefix, last_time)
+    uuid = SecureRandom.uuid
```

Comment:
If this fails half way through, can we end up in a state where half the archives are uploaded, and then the transaction rolls back and when it runs again we get everything uploaded so end up with double the archives on half of the dates?

I haven't thought about it too much, but can we build this `uuid` from some sort of hash of the batch to avoid this situation?

I realise that in most cases, only one days worth of emails will be archived so this won't be a problem though.

Reply:
Yeah this is a problem we have with every upload to S3 - if we think it's failed but it actually succeeded we may end up duplicating the data. The only way I can think around this is having a very predictable batching so that we can generate a reliable hash but this seems difficult to do relative to the risk involved.